pkg/hive/health/metrics.go

```diff
@@ -44,13 +53,24 @@ type metricPublisherParams struct {
 	Metrics  *Metrics
 }
 
-// metricPublisher periodically publishes the hive module health metric (hive_health_status_levels).
+// metricPublisher periodically publishes the hive module health metric
+// * cilium_hive_status
+// * cilium_hive_degraded_status
 func metricPublisher(p metricPublisherParams) {
 	// Performs the actual writing to the metric. Extracted to make testing easy.
-	publish := func(stats map[types.Level]uint64) {
+	publish := func(stats map[types.Level]uint64, degradedModuleCount map[string]uint64) {
 		for l, v := range stats {
 			p.Metrics.HealthStatusGauge.WithLabelValues(strings.ToLower(string(l))).Set(float64(v))
 		}
+		for k, v := range degradedModuleCount {
+			// If the module is healthy attempt to remove any associated metrics with that module
+			if v == 0 {
+				p.Metrics.DegradedHealthStatusGauge.DeleteLabelValues(k)
```

Comment:
One bit of an edge case with this is if all health statuses for with the same moduleID get deleted then it won't get cleaned up here as presumably it won't even show up as a 0 in the map.  

However I don't think this ever actually happens - given this metric isn't on by default, it's probably not worth the extra complexity of making this this will always clean up every module key.

Reply:
```			_, ok := idToStatus[obj.ID.Module.String()]
			if obj.Level == types.LevelDegraded {
				idToStatus[obj.ID.Module.String()]++
			} else if !ok {
				idToStatus[obj.ID.Module.String()] = 0
			}
```

All module ids should show up in the map as 0? I'm not quite following, could you please clarify it a bit?

Reply:
For reference, I discussed this with @ovidiutirla on slack, the specific condition I was getting at here is:

Say we just had one reporter: x.a -> degraded.

* metrics has module 'x' = 1
* x.a tree is removed via [Close](https://github.com/cilium/cilium/blob/724d5b5f3092a0f329d9b36e674adfa020c27188/pkg/hive/health/provider.go#L228-L232).
* x.a is removed [from statedb table](https://github.com/cilium/cilium/blob/724d5b5f3092a0f329d9b36e674adfa020c27188/pkg/hive/health/provider.go#L112)
This triggers the statedb metrics watch channel, causing metrics iteration.
* idToStatus is counted up, but it doesn't contain module x.
* We iterate all module counts in idToStatus, but we fail to remove the metric row... so the metric is left at 1 indefinitely.

Reply:
> One bit of an edge case with this is if all health statuses for with the same moduleID get deleted then it won't get cleaned up here as presumably it won't even show up as a 0 in the map.

Thinking through this a bit more, this might be more likely than I thought, since the Close function deletes all sub-component reports in the module, all it would take is a Close() on a root reporter to cause a leak.

Reply:
@tommyp1ckles, thanks for the discussion we have on Slack. I've updated the PR to also handle your concern. Please have another look before merging this.